linkerd/app/outbound/src/tcp/logical.rs

```diff
@@ -17,40 +50,116 @@ impl<N> Outbound<N> {
             concrete
-                .push_map_target(Concrete::from)
-                .check_new_service::<(ConcreteAddr, Logical), I>()
-                .push(profiles::split::layer())
-                .push_on_service(
-                    svc::layers()
-                        .push(
-                            rt.metrics
-                                .proxy
-                                .stack
-                                .layer(crate::stack_labels("tcp", "logical")),
-                        )
-                        // TODO(ver) We should instead buffer per concrete
-                        // target.
-                        .push_buffer("TCP Logical", tcp_connection_buffer),
-                )
-                // TODO(ver) Can we replace this evicting cache? The detect
-                // stack would have to hold/reuse inner stacks.
-                .push_idle_cache(*discovery_idle_timeout)
-                .check_new_service::<Logical, I>()
-                .instrument(|_: &Logical| debug_span!("tcp"))
-                .check_new_service::<Logical, I>()
+                // Share the concrete stack with each router stack.
+                .push_new_clone()
+                // Rebuild this router stack every time the profile changes.
+                .push_on_service(router)
+                .push(svc::NewSpawnWatch::<Profile, _>::layer_into::<Params>())
                 .push(svc::ArcNewService::layer())
         })
     }
 }
+
+// === impl Params ===
+
+impl From<(Profile, Logical)> for Params {
+    fn from((profile, logical): (Profile, Logical)) -> Self {
+        // Create concrete targets for all of the profile's routes.
+        let (backends, distribution) = if profile.targets.is_empty() {
+            let concrete = super::Concrete {
+                resolve: ConcreteAddr(logical.logical_addr.clone().into()),
+                logical: logical.clone(),
+            };
+            let backends = std::iter::once(concrete.clone()).collect();
+            let distribution = Distribution::first_available(std::iter::once(concrete));
+            (backends, distribution)
+        } else {
+            let backends = profile
+                .targets
+                .iter()
+                .map(|t| super::Concrete {
+                    resolve: ConcreteAddr(t.addr.clone()),
+                    logical: logical.clone(),
+                })
+                .collect();
+            let distribution = Distribution::random_available(profile.targets.iter().cloned().map(
+                |profiles::Target { addr, weight }| {
+                    let concrete = Concrete {
+                        logical: logical.clone(),
+                        resolve: ConcreteAddr(addr),
+                    };
+                    (concrete, weight)
+                },
+            ))
+            .expect("distribution must be valid");
+
+            (backends, distribution)
+        };
```

Comment:
it occurs to me that the code for generating the distribution is identical across the TCP and HTTP routeparams and could probably be shared. not a big deal though.

Reply:
yeah, agreed. let's see where the config goes before we try to unify things. I'm not sure we want a single `Concrete` type, for instance...